hledger-lib/Hledger/Data/DayPartition.hs

```diff
@@ -0,0 +1,275 @@
+{-|
+A partition of time into contiguous spans, for defining reporting periods.
+-}
+module Hledger.Data.DayPartition
+( DayPartition
+, boundariesToDayPartition
+, boundariesToMaybeDayPartition
+
+, lookupDayPartition
+, unionDayPartitions
+
+, dayPartitionToNonEmpty
+, dayPartitionToList
+, dayPartitionSpans
+, dayPartitionToDateSpans
+, dayPartitionToPeriodData
+, maybeDayPartitionToDateSpans
+
+, splitSpan
+, intervalBoundaryBefore
+
+, tests_DayPartition
+) where
+
+import qualified Data.IntMap.Strict as IM
+import Data.List.NonEmpty (NonEmpty(..))
+import qualified Data.List.NonEmpty as NE
+import Data.Time (Day(..), addDays, addGregorianMonthsClip, addGregorianYearsClip, fromGregorian)
+
+import Hledger.Data.Dates
+import Hledger.Data.PeriodData
+import Hledger.Data.Types
+import Hledger.Utils
+
+
+-- | A partition of time into contiguous spans, along with a historical period
+-- before any of the spans.
+--
+-- This is a newtype wrapper around 'PeriodData Day', where the start dates are
+-- the keys and the end dates are the values. Spans are stored in inclusive format
+-- [start, end]. Note that this differs from 'DateSpan' which uses [start, end)
+-- format.
+--
+-- The constructor is not exported so that we can ensure the spans are valid
+-- partitions of time.
+newtype DayPartition = DayPartition { dayPartitionToPeriodData :: PeriodData Day } deriving (Eq, Ord, Show)
+
+-- Developer's note. All constructors must guarantee that:
+--   1. The value stored in pdperiods has at least one key.
+--   2. The value stored in pdpre equals one day before the smallest key in pdperiods.
+--   3. The value stored in each entry of pdperiods equals one day before the
+--      next largest key, except for the value associated to the largest key.
```

Comment:
Who needs to follow these rules - a user of this module ? When using which functions ? Would it be better to mention these as invariants, in DayPartition's doc ?